src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -250,231 +220,94 @@ public void AddPeer(ITxPoolPeer peer)
-                || isTxNonceTooFarInFuture)
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                Metrics.PendingTransactionsTooFarInFuture++;
-                if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, nonce in future.");
-                return AddTxResult.FutureNonce;
-            }
-
-            UInt256 balance = account?.Balance ?? UInt256.Zero;
-            UInt256 payableGasPrice = tx.CalculatePayableGasPrice(spec.IsEip1559Enabled, CurrentBaseFee, balance);
-
-            bool overflow = spec.IsEip1559Enabled && UInt256.AddOverflow(tx.MaxPriorityFeePerGas, tx.MaxFeePerGas, out _);
-            overflow |= UInt256.MultiplyOverflow(payableGasPrice, (UInt256) tx.GasLimit, out UInt256 cost);
-            overflow |= UInt256.AddOverflow(cost, tx.Value, out cost);
-            if (overflow)
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, cost overflow.");
-                return AddTxResult.BalanceOverflow;
-            }
-            else if (balance < cost)
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, insufficient funds.");
-                return AddTxResult.InsufficientFunds;
-            }
-                
-            if (isTxPoolFull
-                && _transactions.TryGetLast(out var lastTx)
-                && payableGasPrice <= lastTx?.GasBottleneck)
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                Metrics.PendingTransactionsTooLowFee++;
-                if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low payable gas price.");
-                return AddTxResult.FeeTooLow;
-            }
-
-            if (managedNonce && CheckOwnTransactionAlreadyUsed(tx, currentNonce))
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                if (_logger.IsTrace)
-                    _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, nonce already used.");
-                return AddTxResult.OwnNonceAlreadyUsed;
-            }
-            
-            return null;
+            return AddCore(tx, startBroadcast);
         }
 
-        private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast, bool isReorg)
+        private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast)
         {
-            bool isInHashCache = _hashCache.Get(tx.Hash);
-            // !!! do not change it to |=
-            bool isKnown = !isReorg && isInHashCache;
-
-            if (!isInHashCache)
-            {
-                _hashCache.Set(tx.Hash);
-            }
```

Comment:
After moving adding to hashCache from the beginning of AddCore to the end (it is in StoreTx right now), now hashCache is misbehaving - if we had tx in _transactions in the past (still is in _txStorage), then we are adding this tx to _transactions, but not adding to hashCache (AddTxResult.AlreadyKnown is returned). We should move adding to hashCache back to the beginning of AddCore or remove _txStorage check. I recommend removing _txStorage at all - it is not useful anymore

Reply:
adding to hash cache is done both at the beginning and at the end - two differen scopes -> block scope and chain scope.
see the interior of the filter

Reply:
tx storage can be removed